File: backend/app/api/v1/gifts.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import get_current_user, get_current_admin
from app.schemas.gift import (
    GiftCreate, GiftUpdate, GiftResponse, GiftFilterParams,
    CategoryCreate, CategoryResponse
)
from app.services.gift_service import GiftService

router = APIRouter(prefix="/gifts", tags=["Gifts"])
category_router = APIRouter(prefix="/categories", tags=["Categories"])


@router.get("/", response_model=list[GiftResponse])
async def list_gifts(
    occasion: str = Query(None),
    relationship: str = Query(None),
    min_price: float = Query(None, ge=0),
    max_price: float = Query(None, ge=0),
    category_id: int = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    """List gifts with optional filtering."""
    params = GiftFilterParams(
        occasion=occasion,
        relationship=relationship,
        min_price=min_price,
        max_price=max_price,
        category_id=category_id,
        skip=skip,
        limit=limit,
    )
    service = GiftService(db)
    return await service.list_gifts(params)


@router.get("/{gift_id}", response_model=GiftResponse)
async def get_gift(gift_id: int, db: AsyncSession = Depends(get_db)):
    """Get a single gift by ID."""
    service = GiftService(db)
    return await service.get_gift(gift_id)


@router.post("/", response_model=GiftResponse, status_code=201)
async def create_gift(
    payload: GiftCreate,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(get_current_admin),
):
    """Create a new gift (admin only)."""
    service = GiftService(db)
    return await service.create_gift(payload)


@router.put("/{gift_id}", response_model=GiftResponse)
async def update_gift(
    gift_id: int,
    payload: GiftUpdate,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(get_current_admin),
):
    """Update an existing gift (admin only)."""
    service = GiftService(db)
    return await service.update_gift(gift_id, payload)


@router.delete("/{gift_id}", status_code=204)
async def delete_gift(
    gift_id: int,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(get_current_admin),
):
    """Delete a gift (admin only)."""
    service = GiftService(db)
    await service.delete_gift(gift_id)


@category_router.get("/", response_model=list[CategoryResponse])
async def list_categories(db: AsyncSession = Depends(get_db)):
    """List all categories."""
    service = GiftService(db)
    return await service.list_categories()


@category_router.post("/", response_model=CategoryResponse, status_code=201)
async def create_category(
    payload: CategoryCreate,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(get_current_admin),
):
    """Create a new category (admin only)."""
    service = GiftService(db)
    return await service.create_category(payload)

File: frontend/src/lib/api/gifts.ts
import { api } from "./client";

export interface Category {
  id: number;
  name: string;
}

export interface Gift {
  id: number;
  title: string;
  description?: string;
  category_id: number;
  category?: Category;
  price: number;
  occasion?: string;
  relationship?: string;
  image_url?: string;
  product_url?: string;
  created_at: string;
}

export interface GiftFilters {
  occasion?: string;
  relationship?: string;
  min_price?: number;
  max_price?: number;
  category_id?: number;
  skip?: number;
  limit?: number;
}

export interface CreateGiftPayload {
  title: string;
  description?: string;
  category_id: number;
  price: number;
  occasion?: string;
  relationship?: string;
  image_url?: string;
  product_url?: string;
}

function buildQS(filters?: GiftFilters): string {
  if (!filters) return "";
  const p = new URLSearchParams();
  if (filters.occasion) p.set("occasion", filters.occasion);
  if (filters.relationship) p.set("relationship", filters.relationship);
  if (filters.min_price != null) p.set("min_price", String(filters.min_price));
  if (filters.max_price != null) p.set("max_price", String(filters.max_price));
  if (filters.category_id != null)
    p.set("category_id", String(filters.category_id));
  if (filters.skip != null) p.set("skip", String(filters.skip));
  if (filters.limit != null) p.set("limit", String(filters.limit));
  return p.toString() ? `?${p}` : "";
}

export const listGifts = (f?: GiftFilters) =>
  api.get<Gift[]>(`/gifts/${buildQS(f)}`);
export const getGift = (id: number) => api.get<Gift>(`/gifts/${id}`);
export const createGift = (p: CreateGiftPayload) =>
  api.post<Gift>("/gifts/", p, true);
export const updateGift = (id: number, p: Partial<CreateGiftPayload>) =>
  api.put<Gift>(`/gifts/${id}`, p, true);
export const deleteGift = (id: number) =>
  api.delete<void>(`/gifts/${id}`, true);

export const listCategories = () => api.get<Category[]>("/categories/");
export const createCategory = (name: string) =>
  api.post<Category>("/categories/", { name }, true);
